File: algorithm-java/src/com/godric/v21_11/algo/C07_BSAwesome.java
package com.godric.v21_11.algo;

/**
 * 局部最小值问题: 在一个先递减后递增的数组中找到极小值点
 */
public class C07_BSAwesome {

    public static int getLessIndex(int[] arr) {

        if (arr == null || arr.length < 1) {
            return -1;
        }

        if (arr.length == 1 || arr[0] < arr[1]) {
            return 0;
        }

        int left = 1;
        int right = arr.length - 2;

        while (left < right) {
            int mid = left + ((right - left) >> 1);
            if (arr[mid] > arr[mid - 1]) {
                right = mid - 1;
            } else if (arr[mid] > arr[mid + 1]) {
                left = mid + 1;
            } else {
                return mid;
            }
        }

        return left;
    }

    public static void main(String[] args) {
        int[] arr = {9, 8, 7, 6, 5, 6, 7, 8, 9};
        System.out.println(arr[getLessIndex(arr)]);
    }

}

File: algorithm-cpp/c07_bs_awesome.cpp
#include<iostream>
#include<vector>
using namespace std;
/**
 * 局部最小值问题: 在一个先递减后递增的数组中找到极小值点
 */
int get_less_index(vector<int> arr)
{
    int left = 0;
    int right = arr.size() - 1;
    while (left <= right)
    {
        int mid = left + ((right - left) >> 1);
        if (arr[mid] < arr[mid-1] && arr[mid] < arr[mid+1])
        {
            return mid;
        }
        if (mid == 0 || arr[mid-1] > arr[mid])
        {
            left = mid + 1;
        } else if (mid == arr.size() - 1 || arr[mid] < arr[mid + 1])
        {
            right = mid - 1;
        }
    }
    return -1;
}


int main()
{
    vector<int> arr = {9, 8, 7, 8, 9, 10, 11};
    cout << get_less_index(arr) << endl;
}